keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1806 @@
+  the pod, either directly or indirectly from the container requests (using the
+  formula in [KEP
+  #753](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/READMEmd#resources-calculation-for-scheduling-and-pod-admission))
+  
+  **Rule: If no requests are provided either at the pod level or at the container level,
+  the pod specification will be considered invalid. This stricter approach
+  encourages users to explicitly defined their resource needs and avoids relying
+  on any arbitrary default values ([arbitrary default minimum
+  values](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/pod_resources.go#L32-L37)
+  that we apply today) that might not be suitable for all scenarios.**
+
+* Specifying Pod Limits: Pod-level limits can be specified in following ways:
+
+  * Explicitly Defined: Users can explicitly set pod-level limits in the pod
+    spec. 
+  * Derived From Containers: If pod-level limit is not set, Kubernetes can
+    derive it from limits set on individual containers within the pod.
+  * Combined Approach: For finer-grained control over individual container
+    resource usage, users can combine pod-level and container-level limits.
+    This approach can help in preventing any single container from monopolizing
+    the resources within the pod. It is particularly useful when you have a
+    critical container that needs guaranteed minimum and maximum resource
+    allocation, while other containers can share the remaining resources more
+    flexibly.
+  * Default to Node Allocatable Resources: If pod-level limit is not set and not
+    all containers within the pod have specified the limits, the pod's limit will
+    default to the allocatable resources of the node where it is scheduled. This
+    ensures the pod can utilize the maximum available resources on the node.
+
+  **Rule: If Pod-level limit is not set, or if it cannot be derived from
+  container-level limits (not all containers have limits set), the containers in
+  the pod can use all the resources allocatable on the node.**
+* The aggregated resource requests of all containers within a pod will never
+  exceed the pod-level resource requests or limits.
+
+  **Rule: Aggregated container requests cannot be greater than pod-level
+  request or limit.**
+* Aggregate container-level limits can be greater than pod-level limits, but the
+  combined resource consumption of all the containers within a pod will always be
+  less than the limit.
+
+  **Rule: Aggregated container limits can be greater than pod-level
+  request or limit, but individual container limit greater than pod-level limit
+  is not allowed.**
+ 
+* **Supported Resources**: Currently, pod level resources support only CPU and
+  memory.
+
+  **Rule: Pod-level resource spec for resources other than CPU and memory is invalid**
+
+##### Comprehensive Tabular View
```

Comment:
Can you explicitly include the design principle we discussed offline here, so that the readers understand the rationale behind. I noticed that you already include most of them below but scattered throughout the document. Centralizing them in a dedicated section would significantly enhance clarity and readability.

1) Prioritize explicit pod-level settings to respect the user's explicit intent and provides clarity
2) Avoid implicitly applying default values at the pod level. 
3) Consistent derivation from container values to ensure consistency and predictability
4) The derivation logic maintains the compatibility with existing behavior

There are several principles we followed during the design, but I failed to spot them from the doc. Maybe we can explicitly state them too:

1) Fairness: The resource allocation mechanism should strive for fairness among containers within a pod, preventing any single container from monopolizing resources unless explicitly configured. For example, OOM score adjustment calculation below demonstrates the fairness in our design. 
2) Flexibility: We carefully consider the flexibility in resource allocation, enabling users to express different resource requirements and sharing strategies as needed. For example, optional  pod-level resources, combination of pod-level and container-level limits, etc. make the resource allocation flexible enough to meet the different requirements.